### PR TITLE
Turn list of txs into tx sequence

### PR DESCRIPTION
**Pull request by aaronbuchwald**
## Why this should be merged

This PR abstracts the load simulator in a different way.

The previous version used a series of workers that performed the tasks of 1) issuing transactions, 2) waiting for the last accepted nonce to exceed the last nonce of the latest transaction, and 3) confirming the transaction receipts in order.

It then created N workers in order to issue and confirm all of the transactions to separate RPC endpoints. This new version switches to using a TxSequence that returns a channel of transactions that need to be issued and confirmed and a worker that plays the role of issuing and confirming transactions.

On top of these two, we can create different issuance strategies to play around with. In this PR, those two different strategies are:
1) issue all transactions at once and confirm them sequentially
2) Issue N transactions, confirm the whole batch, and repeat until iterating through the entire transaction sequence

There are two motivating factors for this change
1) Make it easier to simulate different issuance strategies
2) Simulate more realistic workloads where each tx agent makes a choice based off of the results of the transactions as opposed to issuing all transactions as a fire and forget operation, which is not a realistic use case.

An example of a more realistic pattern is for warp messaging. For the warp messaging PR https://github.com/ava-labs/subnet-evm/pull/586 we would like to perform some degree of stress testing for sending/receiving warp messages.

The goal of this refactor is to make it as easy as possible to construct a realistic load simulation for warp that can perform the following process and scale up to stress the system:

1) Issue a transaction to send a warp message
2) Aggregate a warp signature and create a transaction to deliver the warp message

A fun version of this load test may be to create a very simple smart contract to send a monotonically increasing message back and forth between two different subnets.

## How this was tested

This PR was tested locally with a variety of parameters including:
1) ws and rpc endpoints
2) multiple endpoints
3) single validator and multi-validator

## How is this documented

The existing README for the simulator still documents how to use the load simulator correctly.

### CODE REVIEW

File: cmd/simulator/load/workers.go [hunk 1]
```diff
@@ -1,106 +0,0 @@
...
-	log.Info("Transaction confirmation completed", "confirmationElapsed", endTime.Sub(confirmationStart))
-	totalTxs := 0
-	for _, worker := range wg.workers {
-		totalTxs += len(worker.txs)
-	}
-
-	// We exclude the time for ConfirmAllTransactions because we assume once the nonce is reported, the transaction has been finalized.
-	// For Subnet-EVM, this assumes that unfinalized data is not exposed via API.
-	issueAndConfirmDuration := confirmationStart.Sub(executionStart)
-	log.Info("Simulator completed", "totalTxs", totalTxs, "TPS", float64(totalTxs)/issueAndConfirmDuration.Seconds())
```
Comment (aaronbuchwald) [inline cmd/simulator/load/workers.go:104]: TODO:
- add back these same metrics
- add metrics for individual transaction latency

File: cmd/simulator/txs/agent.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// issueNAgent issues and confirms a batch of N transactions at a time.
+type issueNAgent[T any] struct {
+	sequence TxSequence[T]
+	worker   Worker[T]
+	n        int
+}
+
+func NewIssueNAgent[T any](sequence TxSequence[T], worker Worker[T], n int) Agent[T] {
```
Comment (ceyonur) [inline cmd/simulator/txs/agent.go:72]: do we use this?

File: cmd/simulator/txs/agent.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+func (a issueNAgent[T]) Execute(ctx context.Context) error {
+	txChan := a.sequence.Chan()
+
+	for {
+		var (
+			txs  = make([]T, 0, a.n)
+			tx   T
+			done bool
+		)
+		for i := 0; i < a.n; i++ {
```
Comment (ceyonur) [inline cmd/simulator/txs/agent.go:89]: How does this work for `NewIssueAllAgent` where `n=0`?

File: cmd/simulator/load/tx_generator.go
Comment (ceyonur) [inline cmd/simulator/load/tx_generator.go:1]: Should we move this to txs package?

File: cmd/simulator/load/tx_generator.go [hunk 4]
```diff
@@ -8,15 +8,18 @@ import (
...
 	ethcrypto "github.com/ethereum/go-ethereum/crypto"
 )
 
+var _ txs.TxSequence[*types.Transaction] = (*txSequence)(nil)
+
 type CreateTx func(key *ecdsa.PrivateKey, nonce uint64) (*types.Transaction, error)
 
 // GenerateTxSequence fetches the current nonce of key and calls [generator] [numTxs] times sequentially to generate a sequence of transactions.
-func GenerateTxSequence(ctx context.Context, generator CreateTx, client ethclient.Client, key *ecdsa.PrivateKey, numTxs uint64) ([]*types.Transaction, error) {
+func GenerateTxSequence(ctx context.Context, generator CreateTx, client ethclient.Client, key *ecdsa.PrivateKey, numTxs uint64) (txs.TxSequence[*types.Transaction], error) {
```
Comment (ceyonur) [inline cmd/simulator/load/tx_generator.go:22]: Do we need this to be exist/exported? I think `GenerateTxSequences` is just a generalized version of this.

File: cmd/simulator/load/loader.go [hunk 5]
```diff
@@ -73,10 +82,11 @@ func CreateLoader(ctx context.Context, config config.Config) (*WorkerGroup, erro
 	client := clients[0]
 	chainID, err := client.ChainID(ctx)
 	if err != nil {
-		return nil, fmt.Errorf("failed to fetch chainID: %w", err)
+		return fmt.Errorf("failed to fetch chainID: %w", err)
 	}
 	signer := types.LatestSignerForChainID(chainID)
 
+	log.Info("Creating transaction sequences...")
 	txGenerator := func(key *ecdsa.PrivateKey, nonce uint64) (*types.Transaction, error) {
```
Comment (ceyonur) [inline cmd/simulator/load/loader.go:90]: It seems this txGenerator func can just be a series of options/params for `GenerateTxSequences`

File: cmd/simulator/load/loader.go [hunk 6]
```diff
@@ -96,24 +106,28 @@ func CreateLoader(ctx context.Context, config config.Config) (*WorkerGroup, erro
 	}
 	txSequences, err := GenerateTxSequences(ctx, txGenerator, clients[0], pks, config.TxsPerWorker)
```
Comment (ceyonur) [inline cmd/simulator/load/loader.go:107]: Do you think `Agent` can handle generating tx sequences itself by taking priv keys + numTx?

File: cmd/simulator/load/loader.go [hunk 7]
```diff
@@ -11,40 +11,48 @@ import (
...
 	"github.com/ava-labs/subnet-evm/params"
 	"github.com/ethereum/go-ethereum/common"
 	ethcrypto "github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/log"
+	"golang.org/x/sync/errgroup"
 )
 
 // CreateLoader creates a WorkerGroup from [config] to perform the specified simulation.
-func CreateLoader(ctx context.Context, config config.Config) (*WorkerGroup, error) {
+func ExecuteLoader(ctx context.Context, config config.Config) error {
```
Comment (anusha-ctrl) [inline cmd/simulator/load/loader.go:25]: comment needs to reflect function  name change

File: cmd/simulator/load/worker.go [hunk 8]
```diff
@@ -10,102 +10,75 @@ import (
...
+		// If the is less than what has already been accepted, the transaction is confirmed
+		if txNonce < tw.acceptedNonce {
+			return nil
+		}
+
 		select {
-		case <-newHeads:
+		case <-tw.newHeads:
 		case <-time.After(time.Second):
 		case <-ctx.Done():
```
Comment (anusha-ctrl) [inline cmd/simulator/load/worker.go:64]: curious about this case `case <-time.After(time.Second):`

File: cmd/simulator/txs/agent.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+// Worker defines the interface for issuance and confirmation of transactions.
+// The caller is responsible for calling Close to cleanup resources used by the
+// worker at the end of the simulation.
+type Worker[T any] interface {
+	IssueTx(ctx context.Context, tx T) error
+	ConfirmTx(ctx context.Context, tx T) error
+	Close(ctx context.Context) error
+}
+
```
Comment (anusha-ctrl) [inline cmd/simulator/txs/agent.go:26]: curious why we are using generic types in this file?

File: cmd/simulator/load/worker.go [hunk 10]
```diff
@@ -10,102 +10,75 @@ import (
...
+		acceptedNonce, err := tw.client.NonceAt(ctx, tw.address, nil)
 		if err != nil {
-			log.Warn("failed to get nonce", "err", err)
-		}
-		if currentNonce >= nonce {
-			return nil
-		} else {
-			log.Info("fetched nonce", "awaiting", nonce, "currentNonce", currentNonce)
+			return fmt.Errorf("failed to await tx %s nonce %d: %w", tx.Hash(), txNonce, err)
 		}
```
Comment (anusha-ctrl) [inline cmd/simulator/load/worker.go:73]: can we still keep the awaiting line. It's helpful to the user.

File: cmd/simulator/load/worker.go [hunk 11]
```diff
@@ -10,102 +10,75 @@ import (
...
-// rejected by the network.
-//
-// Assumes that a non-zero number of transactions were already generated and that they were issued
-// by this worker.
-func (w *Worker) AwaitTxs(ctx context.Context) error {
-	nonce := w.txs[len(w.txs)-1].Nonce()
-
+// what do we want
+// ongoing goroutine that updates the nonce and sends/drops an event if there is any listener
+func NewSingleAddressTxWorker(ctx context.Context, client ethclient.Client, address common.Address) *singleAddressTxWorker {
```
Comment (anusha-ctrl) [inline cmd/simulator/load/worker.go:30]: comment needs to be updated

File: cmd/simulator/txs/agent.go [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+		sequence: sequence,
+		worker:   worker,
+	}
+}
+
+func (a issueAllAgent[T]) Execute(ctx context.Context) error {
+	txChan := a.sequence.Chan()
+
+	txs := make([]T, 0, 100)
+	for tx := range txChan {
```
Comment (anusha-ctrl) [inline cmd/simulator/txs/agent.go:49]: are we now limited to 100 txs per run?

Comment (anusha-ctrl) [inline cmd/simulator/txs/agent.go:49]: seems like there are two functions that do the same thing. One has a fixed N at 100 and one has a nonfixed N. We should keep the nonfixed N function and have the user decide N in config cli flags

File: cmd/simulator/load/worker.go (hunk 11 above)
Comment (anusha-ctrl) [reply cmd/simulator/load/worker.go:30]: done

File: cmd/simulator/load/tx_generator.go (hunk 4 above)
Comment (anusha-ctrl) [reply cmd/simulator/load/tx_generator.go:22]: i can unexport this

File: cmd/simulator/load/tx_generator.go
Comment (anusha-ctrl) [reply cmd/simulator/load/tx_generator.go:1]: yes

File: cmd/simulator/load/workers.go (hunk 1 above)
Comment (anusha-ctrl) [reply cmd/simulator/load/workers.go:104]: in the next pr

File: cmd/simulator/txs/agent.go (hunk 12 above)
Comment (anusha-ctrl) [reply cmd/simulator/txs/agent.go]: done

File: cmd/simulator/load/worker.go (hunk 10 above)
Comment (aaronbuchwald) [reply cmd/simulator/load/worker.go:72]: I think it's excessive to do so for every nonce at the info level. Could we make it debug?

**Review by darioush — APPROVED** [review]

File: cmd/simulator/README.md [hunk 13]
```diff
@@ -43,14 +43,14 @@ The staking-enabled flag is only for local testing. Disabling staking serves two
 
 1. Ignore stake weight on the P-Chain and count each connected peer as having a stake weight of 1
 2. Automatically opts in to validate every Subnet
-:::
+   :::
```
[the remark this answers is not in the document]
Comment (anusha-ctrl) [reply cmd/simulator/README.md:46]: my editor added that

File: cmd/simulator/load/worker.go [hunk 14]
```diff
@@ -10,102 +10,75 @@ import (
...
+
+func (tw *singleAddressTxWorker) IssueTx(ctx context.Context, tx *types.Transaction) error {
+	return tw.client.SendTransaction(ctx, tx)
+}
+
+func (tw *singleAddressTxWorker) ConfirmTx(ctx context.Context, tx *types.Transaction) error {
+	txNonce := tx.Nonce()
+
 	for {
+		// If the is less than what has already been accepted, the transaction is confirmed
```
[the remark this answers is not in the document]
Comment (anusha-ctrl) [reply cmd/simulator/load/worker.go:55]: thanks

File: cmd/simulator/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (anusha-ctrl) [reply cmd/simulator/README.md:46]: will fix

File: cmd/simulator/load/tx_generator.go (hunk 4 above)
Comment (anusha-ctrl) [reply cmd/simulator/load/tx_generator.go]: never mind it is used in funder

**Comment by aaronbuchwald** [comment]
LGTM once the `:::` is removed

**Pull request merged**